README.md
# dvsource-v4l2-other

A source for DVswitch which supports any Video4Linux 2 device.

## Quick Start

```
dvswitch --host=localhost --port=2000 &
./dvsource-v4l2-other --device=/dev/video0 --host=localhost --port=2000
```

#### Quick Start with MJPEG Devices (such as HDMI2USB.tv)

Due to limited USB2.0 bandwidth, many Video4Linux devices support both a low frame rate "raw mode" and a high frame rate "mjpeg mode". By default dvsource-v4l2-other will pick the raw mode, to force it to use the mjpeg mode you need to use the `--caps` command line as follows;

```
dvswitch --host=localhost --port=2000 &
./dvsource-v4l2-other --device=/dev/video0 --host=localhost --port=2000 --caps="image/jpeg,width=1280,height=720"
```

## Dependencies

 * gstreamer-1.2.4 (including, base, good and libav plugins)
 * gst-plugins-dvswitch -- https://github.com/timvideos/gst-plugins-dvswitch

### Installing on Debian and Ubuntu

#### Dependencies
The following packages are needed;

 * gstreamer1.0-tools
 * gstreamer1.0-plugins-good
 * gstreamer1.0-plugins-base
 * gstreamer1.0-libav
 * libgstreamer1.0-0
 * gstreamer1.0-dvswitch
   * Download from https://github.com/timvideos/gst-plugins-dvswitch

##### Installing the dependencies
```
sudo apt-get install \
  gstreamer1.0-tools \
  gstreamer1.0-plugins-good \
  gstreamer1.0-plugins-base \
  gstreamer1.0-libav \
  libgstreamer1.0-0
```

##### Installing the gstreamer dvswitch plugin

If your distro has gstreamer1.0-dvswitch
```
sudo apt-get install gstreamer1.0-dvswitch
```

Else, install from the ppa
```
sudo add-apt-repository ppa:aps-sids/ppa
sudo apt-get update
sudo apt-get install gstreamer1.0-dvswitch
```

Else, build the gstreamer1.0-dvswitch package yourself
```
git clone git://github.com/timvideos/gst-plugins-dvswitch.git
cd gst-plugins-dvswitch
dpkg-checkbuilddeps
# Install any missing build dependencies
dpkg-buildpackage -b
dpkg --install ../gstreamer1.0-dvswitch*.deb
```

# Usage

```
usage: dvsource-v4l2-other [-d DEVICE] [-c CAPS] [-s {ntsc,pal}]
                           [-a {4:3,16:9}] [-t TIMEOUT]
                           [-r {48000,44100,32000}]
                           [-n TYPE]
                           [-v] [-x] [-h HOST] [-p PORT] [--help]

optional arguments:
  -d DEVICE, --device DEVICE
                        Video4Linux device to read the input from.
  -c CAPS, --caps CAPS  gstreamer caps to force v4l2src device to use.
  -s {ntsc,pal}, --system {ntsc,pal}
                        Specify the video system to use.This must match the
                        system used by DVswitch.
  -a {4:3,16:9}, --aspect {4:3,16:9}
                        Choose DV output aspect ratio.
  -t TIMEOUT, --timeout TIMEOUT
                        How long to wait when terminating subprocess before
                        killing.
  -r {48000,44100,32000}, --rate {48000,44100,32000}
                        Specify the sample rate, in Hz. **Only 48000kHz is
                        supported.**
  -n TYPE, --fake TYPE
                        Use a fake source rather then a real V4L2 device.
  -v, --verbose         Increase output verbosity
  -x, --display         Display the incoming video locally.
  -h HOST, --host HOST  Specify the network address on which DVswitch is
                        listening. The host address may be specified by name
                        or as an IPv4 or IPv6 literal.
  -p PORT, --port PORT  Specify the network address on which DVswitch is
                        listening. The host address may be specified by name
                        or as an IPv4 or IPv6 literal.
  --help                show this help message and exit
```

### Build man page

Build the man page from argparse arguments using following command.

```
$ python setup.py build_manpage --output=FILENAME --parser=dvsource-v4l2-other:get_parser
```

## TODO

 - [ ] Support for ~/.dvswitchrc


debian/changelog
dvsource-v4l2-other (0.1.1-1) UNRELASED; urgency=low

  * Initial release

 -- Stefano Rivera <stefanor@debian.org>  Mon, 07 Sep 2015 19:05:50 -0700


common.py
#! /usr/bin/python
#
# -*- coding: utf-8 -*-
# vim: set ts=4 sw=4 et sts=4 ai:
#

"""
dvsource-v4l2-other - Video4Linux2 source for DVswitch which supports any device.
"""

import argparse
import atexit
import os
import tempfile
import time
import subprocess
import shutil

subprocess.DEVNULL = file(os.devnull, "rw+")


class dvsource_common(object):

    @staticmethod
    def get_arguments():
        # We have to disable help as we want -h to be used for host.
        parser = argparse.ArgumentParser(add_help=False)
        parser.add_argument(
            "--help", action='help',
            help="show this help message and exit")

        parser.add_argument(
            "-v", "--verbose", action="store_true",
            help="Increase output verbosity")

        # Control the DV format we will send to dvswitch
        parser.add_argument(
            "-f", "--format", default="ntsc", choices=["ntsc", "pal"],
            help="Choose DV output format.")
        parser.add_argument(
            "-a", "--aspect", default="4:3", choices=["4:3", "16:9"],
            help="Choose DV output aspect ratio.")

        parser.add_argument(
            "-x", "--display", action="store_true",
            help="Display the incoming video locally.")

        config = get_dvswitchrc()
        # dvswitch arguments
        parser.add_argument(
            "-h", "--host", 
            help=(""
                "Specify the network address on which DVswitch is listening. The host"
                " address may be specified by name or as an IPv4 or IPv6 literal."),
            default = config.get("MIXER_HOST", None),
            required = config.get("MIXER_HOST", None) == None,
            )
        parser.add_argument(
            "-p", "--port",
            help=(""
                "Specify the network address on which DVswitch is listening. The host"
                " address may be specified by name or as an IPv4 or IPv6 literal."),
            default = config.get("MIXER_PORT", None),
            required = config.get("MIXER_PORT", None) == None,
            )


    @staticmethod
    def parse_dvswitchrc(configfile):
        r = {}
        for line in file(configfile, "r").readlines():
            cmt = line.find('#')
            if cmt > 0:
                line = line[cmt:]

            line = line.strip()
            if not line:
                continue

            key, value = line.split('=', 1)
            r[key] = value
        return r

    @staticmethod
    def get_dvswitchrc():
        import shlex
        from os import path
        configs = [path.expanduser("~"), ".", "/etc"]

        actual_config = {}
        for dirname in configs:
            configfile = path.join(dirname, ".dvswitchrc")
            if path.exists(configfile):
                actual_config.update(parse_dvswitchrc(configfile))

        if actual_config.get("MIXER_HOST", None) == "0.0.0.0":
            actual_config["MIXER_HOST"] = "127.0.0.1"

        return actual_config

    ###############################################################################
    # Code to check dependencies
    ###############################################################################
    @staticmethod
    def check_command(name):
        try:
            output = subprocess.check_output(["which", name])
            if args.verbose:
              print "Using", name, "found at", output.strip()
        except subprocess.CalledProcessError, e:
            print "Unable to find required command:", name
            raise

    @staticmethod
    def check_gst_module(name):
        try:
            subprocess.check_call(["gst-inspect-0.10", name], stdout=subprocess.DEVNULL)
        except subprocess.CalledProcessError, e:
            print "Unable to find required gstreamer module", name
            raise

###############################################################################
# Code to which actually does the work
###############################################################################

    def get_gstreamer_pipeline(args):
        cmd = (
            # Allow at most 1ms of data to be in the buffer, 2=GST_QUEUE_LEAK_DOWNSTREAM
            "queue leaky=downstream max-size-buffers=1 ! " +
            " " +
            # Convert to 4:3 format by adding borders if needed
            {"4:3":
                # Convert to 4:3 format by adding borders if needed
                "videoscale add-borders=1 ! video/x-raw-yuv,width=1024,height=768,pixel-aspect-ratio=\(fraction\)1/1 !",
             "16:9":
                # Convert to 16:9 format by adding borders if needed
                "videoscale add-borders=1 ! video/x-raw-yuv,width=1280,height=720,pixel-aspect-ratio=\(fraction\)1/1 !",
            }[args.aspect] +
            " " +
            {"ntsc-4:3": 
                # Convert to 4:3 with non-square pixels
                "videoscale ! video/x-raw-yuv,width=720,height=480,pixel-aspect-ratio=\(fraction\)10/11 !",
             "ntsc-16:9": 
                # Convert to 4:3 with non-square pixels
                "videoscale ! video/x-raw-yuv,width=720,height=480,pixel-aspect-ratio=\(fraction\)40/33 !",
             "pal-4:3":
                # Convert to 4:3 with non-square pixels
                "videoscale ! video/x-raw-yuv,width=720,height=576,pixel-aspect-ratio=\(fraction\)59/54 !",
             "pal-4:3":
                # Convert to 4:3 with non-square pixels
                "videoscale ! video/x-raw-yuv,width=720,height=576,pixel-aspect-ratio=\(fraction\)118/81 !",
            }["%s-%s" % (args.format, args.aspect)] +
            " " +
            {"ntsc": 
                # Convert the framerate to 30fps
                "videorate ! video/x-raw-yuv,framerate=\(fraction\)30000/1001 !",
             "pal":
                # Convert the framerate to 25fps
                "videorate ! video/x-raw-yuv,framerate=\(fraction\)25/1 !",
            }[args.format] +
            " " +
            ["", "tee name=t ! "][args.display] +
            " " +
            "queue leaky=downstream max-size-buffers=1 ! " +
            " " +
            # Convert to DV format
            "ffmpegcolorspace ! ffenc_dvvideo ! ffmux_dv !" +
            " " +
            # Output to the fifo
            "dvswitchsink host=%s port=%s" % (args.host, args.port) +
            " " +
            ["", "t.! queue max-size-buffers=1 leaky=downstream ! ffmpegcolorspace ! xvimagesink"][args.display]
            )

###############################################################################
# Main function
###############################################################################
def main():
    # Check that gstreamer stuff is installed
    check_command("gst-inspect-0.10")
    check_command("gst-launch-0.10")
    check_gst_module("v4l2src")
    check_gst_module("decodebin2")
    check_gst_module("videoscale")
    check_gst_module("videorate")
    check_gst_module("ffmpegcolorspace")
    check_gst_module("ffenc_dvvideo")
    check_gst_module("ffmux_dv")
    check_gst_module("filesink")
    check_gst_module("dvswitchsink")

    # Launch the sub-commands
    gst = launch_gstreamer()

    try:
        while True:
            if gst.poll() != None:
                raise OSError("gst-launch command terminated!")

            # FIXME: Add some type of monitoring of CPU usage here...
            if args.verbose:
                print "gst-launch happily running!"

            time.sleep(1.0)
    finally:
        exitstart = time.time()

        # To not get errors, terminate gst first, then dvsource

        try:
            if args.verbose:
                print "Terminating gst-launch"
            gst.terminate()
        except Exception, e:
            print "Error terminating gst-launch", e

        while True:
            if gst.poll() != None:
                break

            if args.verbose:
                print "Waiting for gst-launch to terminate"
            time.sleep(1)

            if time.time() - exitstart > args.timeout:
                print "Timeout waiting for gst-launch",
                print "to terminate, killing."

		try:
		    gst.kill()
		except Exception, e:
		    print "Error killing gst-launch", e

args = None
if __name__ == "__main__":
    args = parser.parse_args()
    main()


build_manpage.py
# -*- coding: utf-8 -*-

"""build_manpage command -- Generate man page from setup()"""

import datetime
from distutils.command.build import build
from distutils.core import Command
from distutils.errors import DistutilsOptionError
import argparse


class build_manpage(Command):

    description = 'Generate man page from setup().'

    user_options = [
        ('output=', 'O', 'output file'),
        ('parser=', None, 'module path to argparser (e.g. mymod:func'),
        ]

    def initialize_options(self):
        self.output = None
        self.parser = None

    def finalize_options(self):
        if self.output is None:
            raise DistutilsOptionError('\'output\' option is required')
        if self.parser is None:
            raise DistutilsOptionError('\'parser\' option is required')
        mod_name, func_name = self.parser.split(':')
        fromlist = mod_name.split('.')
        try:
            mod = __import__(mod_name, fromlist=fromlist)
            self._parser = getattr(mod, func_name)()
        except ImportError, err:
            raise
        self._parser.formatter = ManPageFormatter()
        self.announce('Writing man page %s' % self.output)
        self._today = datetime.date.today()

    def _markup(self, txt):
        return txt.replace('-', '\\-')

    def _write_header(self):
        appname = self.distribution.get_name()
        ret = []
        ret.append('.TH %s 1 %s\n' % (self._markup(appname),
                                      self._today.strftime('%Y\\-%m\\-%d')))
        description = self.distribution.get_description()
        if description:
            name = self._markup('%s - %s' % (appname,
                                             description.splitlines()[0]))
        else:
            name = self._markup(appname)
        ret.append('.SH NAME\n%s\n' % name)
        synopsis = self._parser.format_usage()
        if synopsis:
            synopsis = synopsis.replace('%s ' % appname, '')
            ret.append('.SH SYNOPSIS\n.B %s\n%s\n' % (self._markup(appname),
                                                      synopsis))
        long_desc = self.distribution.get_long_description()
        if long_desc and long_desc != "UNKNOWN":
            ret.append('.SH DESCRIPTION\n%s\n' % self._markup(long_desc))
        return ''.join(ret)

    def _write_options(self):
        ret = ['.SH OPTIONS\n']
        ret.append(self._parser.format_help())
        return ''.join(ret)

    def _write_footer(self):
        ret = []
        appname = self.distribution.get_name()
        author = '%s <%s>' % (self.distribution.get_author(),
                              self.distribution.get_author_email())
        ret.append(('.SH AUTHORS\n.B %s\nwas written by %s.\n'
                    % (self._markup(appname), self._markup(author))))
        homepage = self.distribution.get_url()
        ret.append(('.SH DISTRIBUTION\nThe latest version of %s may '
                    'be downloaded from\n'
                    '.UR %s\n.UE\n'
                    % (self._markup(appname), self._markup(homepage),)))
        return ''.join(ret)

    def run(self):
        manpage = []
        manpage.append(self._write_header())
        manpage.append(self._write_options())
        manpage.append(self._write_footer())
        stream = open(self.output, 'w')
        stream.write(''.join(manpage))
        stream.close()


class ManPageFormatter(argparse.HelpFormatter):

    def __init__(self,
                 indent_increment=2,
                 max_help_position=24,
                 width=None):
        argparse.HelpFormatter.__init__(
            self, indent_increment, max_help_position, width)

    def _markup(self, txt):
        return txt.replace('-', '\\-')

    def format_usage(self, usage):
        return self._markup(usage)

    def format_heading(self, heading):
        if self.level == 0:
            return ''
        return '.TP\n%s\n' % self._markup(heading.upper())

    def format_option(self, option):
        result = []
        opts = self.option_strings[option]
        result.append('.TP\n.B %s\n' % self._markup(opts))
        if option.help:
            help_text = '%s\n' % self._markup(self.expand_default(option))
            result.append(help_text)
        return ''.join(result)


build.sub_commands.append(('build_manpage', None))



dvsource-v4l2-other.py
#! /usr/bin/python
#
# -*- coding: utf-8 -*-
# vim: set ts=4 sw=4 et sts=4 ai:
#

"""
dvsource-v4l2-other - Video4Linux2 source for DVswitch which supports any device.
"""

import argparse
import os
import time
import subprocess

subprocess.DEVNULL = file(os.devnull, "rw+")


###############################################################################
# Argument parsing
###############################################################################

parser = argparse.ArgumentParser(prog='dvsource-v4l2-other', add_help=False)
parser.add_argument(
    "-d", "--device", default="/dev/video0",
    help="Video4Linux device to read the input from.")
parser.add_argument(
    "-c", "--caps", default="",
    help="gstreamer caps to force v4l2src device to use.")

parser.add_argument(
    "-s", "--system", default="pal", choices=["ntsc", "pal"],
    help="Specify the video system to use.This must match the system used by DVswitch.")
parser.add_argument(
    "-a", "--aspect", default="4:3", choices=["4:3", "16:9"],
    help="Choose DV output aspect ratio.")
parser.add_argument(
    "-t", "--timeout", type=int, default=10,
    help="How long to wait when terminating subprocess before killing.")

parser.add_argument(
    "-r", "--rate", default="48000", choices=["48000", "44100", "32000"],
    help="Specify the sample rate, in Hz. **Only 48000kHz is supported.**")


fake_types = {
    None: 0,
    "": 0,
    "smpte": 0,
    "snow": 1,
    "black": 2,
    "white": 3,
    "red": 4,
    "green": 5,
    "blue": 6,
    "checkers-1": 7,
    "checkers-2": 8,
    "checkers-4": 9,
    "checkers-8": 10,
    "circular": 11,
    "blink": 12,
    "smpte": 13,
    "zone-plate": 14,
    "gamut": 15,
    "chroma-zone-plate": 16,
    "solid-color": 17,
    "ball": 18,
    "smpte": 19,
    "bar": 20,
    }
parser.add_argument(
    "-n", "--fake", choices=fake_types.keys(),
    help="Use a fake source rather then a real V4L2 device.")

parser.add_argument(
    "-v", "--verbose", action="store_true",
    help="Increase output verbosity")

parser.add_argument(
    "-x", "--display", action="store_true",
    help="Display the incoming video locally.")

###############################################################################
# dvswitch arguments and .dvswitchrc parsing
###############################################################################

def parse_dvswitchrc(configfile):
    r = {}
    for line in file(configfile, "r").readlines():
        cmt = line.find('#')
        if cmt > 0:
            line = line[cmt:]

        line = line.strip()
        if not line:
            continue

        key, value = line.split('=', 1)
        r[key] = value
    return r


def get_dvswitchrc():
    configs = [os.path.expanduser("~"), ".", "/etc"]

    actual_config = {}
    for dirname in configs:
        configfile = os.path.join(dirname, ".dvswitchrc")
        if os.path.exists(configfile):
            actual_config.update(parse_dvswitchrc(configfile))

    if actual_config.get("MIXER_HOST", None) == "0.0.0.0":
        actual_config["MIXER_HOST"] = "127.0.0.1"

    return actual_config

config = get_dvswitchrc()
# dvswitch arguments
parser.add_argument(
    "-h", "--host", 
    help=(""
        "Specify the network address on which DVswitch is listening. The host"
        " address may be specified by name or as an IPv4 or IPv6 literal."),
    default = config.get("MIXER_HOST", None),
    required = config.get("MIXER_HOST", None) == None,
    )
parser.add_argument(
    "-p", "--port",
    help=(""
        "Specify the network address on which DVswitch is listening. The host"
        " address may be specified by name or as an IPv4 or IPv6 literal."),
    default = config.get("MIXER_PORT", None),
    required = config.get("MIXER_PORT", None) == None,
    )

###############################################################################

parser.add_argument(
    "--help", action='help',
    help="show this help message and exit")

###############################################################################
# Code to check dependencies
###############################################################################
def check_command(name, package=None):
    try:
        output = subprocess.check_output(["which", name])
        if args.verbose:
          print "Using", name, "found at", output.strip()
    except subprocess.CalledProcessError, e:
        print "Unable to find required command:", name
        if package:
            print "Please try 'sudo apt-get install %s'" % package
        raise

def check_gst_module(name, package=None, extra_help=None):
    try:
        subprocess.check_call(["gst-inspect-1.0", name], stdout=subprocess.DEVNULL)
    except subprocess.CalledProcessError, e:
        print "Unable to find required gstreamer module", name
        if package:
            print "Please try 'sudo apt-get install %s'" % package
        if extra_help:
            print extra_help
        raise

###############################################################################
# Code to which actually does the work
###############################################################################

def launch_gstreamer():
    if args.caps:
        args.caps += " ! "

    cmd = ("gst-launch-1.0" +
        " " +
        {True: "-v",
         False: ""
        }[args.verbose] +
        " " +
        # Video Pipeline --------------------
        # -----------------------------------
        {True:
            # Read the v4l2 input and decode it if it's a mjpeg input
            "v4l2src device=%s ! " % args.device +
            args.caps +
            "decodebin ! ",
         False:
          "videotestsrc is-live=true pattern=%s !" % fake_types[args.fake],
        }[args.fake == None] +
        " " +
        # Allow at most 1ms of data to be in the buffer, 2=GST_QUEUE_LEAK_DOWNSTREAM
        "queue leaky=downstream max-size-buffers=1 ! " +
        " " +
        # Convert to 4:3 format by adding borders if needed
        {"4:3":
            # Convert to 4:3 format by adding borders if needed
            "videoscale add-borders=1 ! video/x-raw,width=1024,height=768,pixel-aspect-ratio=\(fraction\)1/1 !",
         "16:9":
            # Convert to 16:9 format by adding borders if needed
            "videoscale add-borders=1 ! video/x-raw,width=1280,height=720,pixel-aspect-ratio=\(fraction\)1/1 !",
        }[args.aspect] +
        " " +
        # Pixel aspect ratio conversion is a bit hairy, see the following links for more information.
        # http://forum.doom9.org/showthread.php?t=111102
        # http://www.sciencemedianetwork.org/wiki/Tutorials/Video/Pixel_Aspect_Ratio
        {"ntsc-4:3": 
            # Convert to 4:3 with non-square pixels (was 10.0/11 ~= 0.91, now 8.0/9 ~= 0.88)
            "videoscale ! video/x-raw,width=720,height=480,pixel-aspect-ratio=\(fraction\)8/9 !",
         "ntsc-16:9": 
            # Convert to 16:9 with non-square pixels (was 40.0/33 ~= 1.21, now 32.0/27 ~= 1.19)
            "videoscale ! video/x-raw,width=720,height=480,pixel-aspect-ratio=\(fraction\)32/27 !",
         "pal-4:3":
            # Convert to 4:3 with non-square pixels (was 59.0/54 ~= 1.09 == ITU-PAR, now 16.0/15 ~= 1.07 == NLE-PAR used by Final Cut / Adobe)
            # I420 == works, Y41B == works but has artifacts, Y42B == not working
            "videoscale ! video/x-raw,width=720,height=576,pixel-aspect-ratio=\(fraction\)16/15 !",
         "pal-16:9":
            # Convert to 16:9 with non-square pixels (was 118.0/81 ~= 1.46 == anamorphic 'ITU', now 64.0/45 ~= 1.42 == anamorphic 'NLE')
            "videoscale ! video/x-raw,width=720,height=576,pixel-aspect-ratio=\(fraction\)64/45 !",
        }["%s-%s" % (args.system, args.aspect)] +
        " " +
        {"ntsc": 
            # Convert the framerate to 30fps
            "videorate ! video/x-raw,framerate=\(fraction\)30000/1001 !",
         "pal":
            # Convert the framerate to 25fps
            "videorate ! video/x-raw,framerate=\(fraction\)25/1 !",
        }[args.system] +
        " " +
        # Convert to color space needed by dvswitch
        # For PAL, I420 works and Y41B works but has artifacts.
        # For NTSC, only Y41B seems to work.
        # Y42B doesn't seem supported by dvswitch at all.
        "videoconvert ! video/x-raw,format=\(string\)" +
        {
            "ntsc": "Y41B",
            "pal": "I420",
        }[args.system] +
        " ! " +
        ["", "tee name=t ! "][args.display] +
        " " +
        "queue leaky=downstream max-size-buffers=1 ! " +
        " " +
        # Convert to DV format
        "videoconvert ! avenc_dvvideo ! avmux_dv name=dvmux !" +
        " " +
        # Output to dvswitch
        "dvswitchsink host=%s port=%s" % (args.host, args.port) +
        " " +
        # -----------------------------------
        # Audio Pipeline --------------------
        # -----------------------------------
        " " +
        # Generate a dummy audio signal
        # 2 channels, 16-bit Linear PCM at 48 kHz
        # 2 channels, 16-bit Linear PCM at 44.1 kHz
        # 4 channels, 12-bit nonlinear PCM channels at 32 kHz (Not supported - gstreamer doesn't support 12-bit nonlinear)
        "audiotestsrc is-live=true wave=sine freq=200 ! audio/x-raw,channels=2,rate=%s,depth=16 ! queue ! dvmux." % args.rate +
        " " +
        # -----------------------------------
        # Local Display ---------------------
        # -----------------------------------
        " " +
        ["", "t. ! queue max-size-buffers=1 leaky=downstream ! videoconvert ! xvimagesink"][args.display]
        )

    cmdargs = {}
    if args.verbose:
        print "Running the gstreamer conversion command of"
        print "   ", cmd
    else:
        cmdargs["stdout"] = subprocess.DEVNULL

    return subprocess.Popen(cmd, shell=True, **cmdargs)


###############################################################################
# Main function
###############################################################################
def main():
    # Check that gstreamer cmd line tools are installed
    check_command("gst-inspect-1.0", "gstreamer1.0-tools")
    check_command("gst-launch-1.0", "gstreamer1.0-tools")
    # Check if the gstreamer modules are installed
    check_gst_module("v4l2src", "gstreamer1.0-plugins-good")
    check_gst_module("decodebin", "gstreamer1.0-plugins-base")

    check_gst_module("videotestsrc", "gstreamer1.0-plugins-base")

    check_gst_module("videoscale", "gstreamer1.0-plugins-base")
    check_gst_module("videorate", "gstreamer1.0-plugins-base")

    check_gst_module("queue", "libgstreamer1.0-0")
    check_gst_module("tee", "libgstreamer1.0-0")

    check_gst_module("videoconvert", "gstreamer1.0-libav")
    check_gst_module("avenc_dvvideo", "gstreamer1.0-libav")
    check_gst_module("avmux_dv", "gstreamer1.0-libav")
    check_gst_module("dvswitchsink", "gstreamer1.0-dvswitch", """
If your distro doesn't ship the gstreamer1.0-dvswitch package you
can find out more information in the README.md file or at
https://github.com/timvideos/dvsource-v4l2-other#installing-the-gstreamer-dvswitch-plugin
""")

    check_gst_module("audiotestsrc", "gstreamer1.0-plugins-base")

    check_gst_module("xvimagesink", "gstreamer1.0-plugins-base")

    # Check the input arguments make sense.
    if args.rate != "48000":
        raise SystemError("Only a --rate of 48000 is supported.")

    # Launch the sub-commands
    gst = launch_gstreamer()

    try:
        while True:
            if gst.poll() != None:
                raise OSError("gst-launch command terminated!")

            # FIXME: Add some type of monitoring of CPU usage here...
            if args.verbose:
                print "gst-launch happily running!"

            time.sleep(1.0)
    except KeyboardInterrupt, e:
        pass
    finally:
        exitstart = time.time()

        try:
            if args.verbose:
                print "Terminating gst-launch"
            gst.terminate()
        except Exception, e:
            print "Error terminating gst-launch", e

        while True:
            if gst.poll() != None:
                break

            if args.verbose:
                print "Waiting for gst-launch to terminate"
            time.sleep(1)

            if time.time() - exitstart > args.timeout:
                print "Timeout waiting for gst-launch",
                print "to terminate, killing."

		try:
		    gst.kill()
		except Exception, e:
		    print "Error killing gst-launch", e


def get_parser():
    return parser


args = None
if __name__ == "__main__":
    args = parser.parse_args()
    main()
